handle_actions.go

```diff
@@ -33,68 +33,34 @@ func (m *Mock) DoActions(c *Context) error {
 }
 
 func (m *Mock) doAction(c *Context, a Action) (err error) {
-	var action string
-
+	actualAction := a.GetActualAction()
 	defer func() {
 		if err != nil {
 			logrus.WithFields(logrus.Fields{
 				"err":    err,
-				"action": action,
+				"action": reflect.TypeOf(actualAction).String(),
 			}).Errorf("failed to do action")
 		}
 	}()
-
-	if err := m.doActionSleep(c, a); err != nil {
-		action = "sleep"
-		return err
-	}
-	if err := m.doActionRedis(c, a); err != nil {
-		action = "redis"
-		return err
-	}
-	if err := m.doActionPublishKafka(c, a); err != nil {
-		action = "publish_kafka"
-		return err
-	}
-	if err := m.doActionPublishAMQP(c, a); err != nil {
-		action = "publish_amqp"
-		return err
-	}
-	if err := m.doActionSendHTTP(c, a); err != nil {
-		action = "send_http"
-		return err
-	}
-
-	if err := m.doActionReplyHTTP(c, a); err != nil {
-		action = "reply_http"
-		return err
-	}
-
-	return nil
+	return actualAction.Perform(*c)
 }
 
-func (m *Mock) doActionSendHTTP(c *Context, a Action) error {
-	if structs.IsZero(a.ActionSendHTTP) {
-		return nil
-	}
-
-	sendHTTP := a.ActionSendHTTP
-
-	bodyStr, err := c.Render(sendHTTP.Body)
+func (actionSendHTTP ActionSendHTTP) Perform(context Context) error {
```

Comment:
nit, for `context` go prefers `ctx` as the variable name. `context` is mostly used as the package name. See https://blog.golang.org/context

Reply:
should we then also rename our `Context` struct, if that's the confusion we're trying to avoid?

Reply:
`Context` is fine, it means a struct; `context` is usually used to refer to the package name from `import "context"`

Reply:
Context should be ok here, implying the package